pyro/infer/reparam/stable.py

```diff
@@ -0,0 +1,198 @@
+    Approximate Levy-Ito decomposition of
+    :class:`~pyro.distributions.StableHMM` random variables whose
+    ``initial_dist`` and ``observation_dist`` are symmetric (but whose
+    ``transition_dist`` may be skewed).
+
+    This is useful for training the parameters of a
+    :class:`~pyro.distributions.StableHMM` distribution, whose
+    :meth:`~pyro.distributions.StableHMM.log_prob` method is undefined.
+
+    This introduces auxiliary random variables conditioned on which the process
+    becomes a :class:`~pyro.distributions.GaussianHMM` . The initial and
+    observation distributions are reparameterized by
+    :class:`SymmetricStableReparam` . The latent transition process is Levy-Ito
+    decomposed into drift + Brownian motion + a compound Poisson process (see
+    [1] section 1.2.6). We neglect the generalized compensated Poisson process
+    of small shocks, and approximate the compound Poisson process with a fixed
+    ``num_shocks`` over the time interval of interest. As ``num_shocks``
+    increases, the shock size cutoffs tend to zero and approximation improves.
+
+    [1] Andreas E. Kyprianou (2013)
+        "Fluctuations of Levy Processes with Applications"
+        https://people.bath.ac.uk/ak257/book2e/book2e.pdf
+
+    :param int num_shocks: Fixed number of shocks to approximate the compound
+        Poisson process component of the Levy-Ito decomposition of latent
+        state. This automatically determines the minimum shock size cutoffs.
+    """
+    def __init__(self, num_shocks):
+        assert isinstance(num_shocks, int) and num_shocks > 0
+        self.num_shocks = num_shocks
+
+    def __call__(self, name, fn, obs):
+        assert isinstance(fn, dist.StableHMM)
+        hidden_dim = fn.hidden_dim
+        duration = fn.transition_dist.batch_shape[-1]
+        stability = fn.initial_dist.base_dist.stability[..., 0]
+
+        # Sample positive and negative shocks to latent state,
+        # independent over each hidden dim.
+        shock_times = pyro.sample("{}_shock_times".format(name),
+                                  dist.Uniform(0, duration)
+                                      .expand([2, self.num_shocks, hidden_dim])
+                                      .to_event(3))
+        shock_sizes = pyro.sample("{}_shock_sizes".format(name),
+                                  dist.Pareto(1, stability.unsqueeze(-1).unsqueeze(-1).unsqueeze(-1))
+                                      .expand([2, self.num_shocks, hidden_dim])
+                                      .to_event(3))
+        pos_shock_times, neg_shock_times = shock_times.unbind(-3)
+        pos_shock_sizes, neg_shock_sizes = shock_sizes.unbind(-3)
+        pos_shocks = _densify_shocks(duration, pos_shock_times, pos_shock_sizes)
+        neg_shocks = _densify_shocks(duration, neg_shock_times, neg_shock_sizes)
+        # FIXME correctly scale pos_shocks and neg_shocks
+        shocks = pos_shocks - neg_shocks
+        # FIXME correct scale and loc
```

Comment:
I think that I understand your purpose here, but I can't find "shocks"-related references to verify the corresponding math. Section 1.2.6 of https://people.bath.ac.uk/ak257/book2e/book2e.pdf gives me no hint. :(

Reply:
For alternate explanations, see
- [wikipedia](https://en.wikipedia.org/wiki/L%C3%A9vy_process#L%C3%A9vy%E2%80%93It%C3%B4_decomposition)
- Theorem 4 in [Barndorff-Nielsen & Shepard](https://pdfs.semanticscholar.org/fe80/07ba98fafa23ddca98ac5d9b1cf1732f70bd.pdf)
- Jim Pitman's [lecture notes](https://www.stat.berkeley.edu/~pitman/s205s03/levy.pdf) which follow Kallenberg closely but in more readable syntax.

By "shocks", I mean the compound Poisson process component of the Levy-Ito representation. This is defined by $\Pi(dx)$ in the linked [Kyprianou (2013)](https://people.bath.ac.uk/ak257/book2e/book2e.pdf) and wikipedia, defined by (N, v(dy)) in Barndorff-Nielsen, and $\nu$ in Pitman's notes. Note the Levy-Ito decomposition of stochastic processes is closely related to the Levy-Khintchine theorem about characteristic functions of those processes. It seems like applied finance folks use the term "shocks", whereas probabilists use the term "jumps"; I can switch to "jumps" if you think that's clearer.

Reply:
OK, I have renamed "shocks" to "jumps" to be more standard.

Reply:
I have removed the Levy-Ito decomposition for now, and will implement it in a subsequent PR.